apisix/plugins/auth-casdoor.lua

```diff
@@ -0,0 +1,138 @@
+local log = core.log
+local ngx = ngx
+
+local plugin_name = "auth-casdoor"
+local schema = {
+    type = "object",
+    properties = {
+        casdoor_endpoint = {type = "string"},
+        client_id = {type = "string"},
+        client_secret = {type = "string"},
+        callback_url = {type = "string"},
+        test = {type = "boolean", default = false}
+    },
+    required = {
+        "callback_url", "casdoor_endpoint", "client_id", "client_secret"
+    }
+}
+
+local _M = {version = 0.1, priority = 2559, name = plugin_name, schema = schema}
+
+local function get_path(uri)
+    local without_query = uri:match("(.-)%?") or uri
+    return without_query:match(".-//[^/]+(/.*)") or without_query
+end
+
+local function fetch_access_token(ctx, conf)
+    local args = core.request.get_uri_args(ctx)
+    if not args or not args.code or not args.state then
+        log.err()
+        return nil, "failed when accessing token. Invalid code or state"
+    end
+    local client = http.new()
+    local url = get_path(conf.casdoor_endpoint) ..
+                    "/api/login/oauth/access_token"
+
+    if conf.test then
+        url = get_path(conf.casdoor_endpoint) ..
+                  "/casdoor_fake_access_token_api"
+    end
+
+    local res, err = client:request_uri(url, {
+        method = "POST",
+        query = {
+            code = args.code,
+            grant_type = "authorization_code",
+            client_id = conf.client_id,
+            client_secret = conf.client_secret
+        }
+    })
+    if not res then return nil, err end
+    local data, err = cjson.decode(res.body)
+
+    if err or not data then return nil, err end
+
+    if not data.access_token then
+        return nil, "failed when accessing token: no access_token contained"
+    end
+
+    return data.access_token, nil
+end
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then return false, err end
+    -- check whether contains extra ? or /
+    local has_extra = string.find(conf.callback_url, "?", 1)
+    if has_extra then
+        return false,
+               "callback_url should not contain get parameters or end up with /"
+    end
+
+    return true, nil
+end
+
+function _M.access(conf, ctx)
+    -- log.info("hit auth-casdoor access")
+    local current_uri = ctx.var.uri
+    local session_obj_read = session.open()
+    -- step 1: check whether hits the callback
+    if get_path(current_uri) == get_path(conf.callback_url) then
+        local access_token, err = fetch_access_token(ctx, conf)
+        if access_token then
+            local original_url = session_obj_read.data.original_uri
+            if not original_url then
+                return 503, "no original_url found in session"
+            end
+
+            local session_obj_write = session.start()
+            session_obj_write.data.access_token = access_token
+            session_obj_write:save()
+            ngx.redirect(original_url, 302)
```

Comment:
Would be better to use this way:
https://github.com/apache/apisix/blob/acf7a0cc7f05aa15de21eb1aac798566f45bcb66/apisix/plugins/redirect.lua#L215-L216

So we can have a common place to handle redirection, as we don't use `ngx.exit` directly.